File: TurnBased-Mess/ProyectoParejasPOO/ProyectoParejasPOO/Enemy.cs
using System;
using System.Collections.Generic;
using System.Security.AccessControl;
using System.Text;

namespace ProyectoParejasPOO
{
    public class Enemy : Character
    {
        public int expOnDeath;
        public int hpScaleFactor;
        public int attackScaleFactor;
        
        // Evento que se dispara al morir el enemigo, para notificar al GameManager y otorgar experiencia al jugador
        // Action<int> es un delegado que representa un método que recibe un int (la experiencia otorgada) y no devuelve nada. El GameManager se suscribirá a este evento para recibir la experiencia cuando el enemigo muera.
        public Action<int> OnDeath;

        public Enemy(string name, int level) : base(name)
        {
            this.name = name;
            this.level = level;

            actions.Add(new SingleTargetAttack());
        }
        public override CharacterAction ChooseAction()
        {
            Random ran = new Random();
            int index = ran.Next(actions.Count);

            CharacterAction action = actions[index];
            action.user = this;
            return action;

        }

        public void ScaleStats()
        {
            maxHealth += hpScaleFactor * level;
            health = maxHealth;
            atk += attackScaleFactor * level;
            defense += Math.Min(0, level - 2);
            expOnDeath += (expOnDeath * level) - level*2;
            /*Console.ForegroundColor = ConsoleColor.Yellow;
            Console.WriteLine($"Creado {name} de nivel {level}, maxHP = {maxHP}");
            Console.ResetColor();*/
        }

    }
    public class Slime : Enemy
    {
        public Slime(int level) : base("Slime", level)
        {
            hpScaleFactor = 3;
            attackScaleFactor = 1;
            expOnDeath = 10;
            speed = 3;
            defense = 0;
            ScaleStats();
        }
    }
    public class Orc : Enemy
    {
        public Orc(int level) : base("Orco", level)
        {
            hpScaleFactor = 4;
            attackScaleFactor = 2;
            expOnDeath = 20;
            speed = 2;
            defense = 0;
            ScaleStats();
        }
    }
    public class Troll : Enemy
    {
        public Troll(int level) : base("Troll", level)
        {
            hpScaleFactor = 5;
            attackScaleFactor = 1;
            expOnDeath = 30;
            speed = 6;
            defense = 1;
            ScaleStats();
        }
    }
    public class MotherSlime : Enemy
    {
        public MotherSlime(int level) : base("Madre Slime", level)
        {
            hpScaleFactor = 8;
            attackScaleFactor = 2;
            expOnDeath = 50;
            speed = 1;
            defense = 2;
            ScaleStats();
        }
    }
    public class Giant : Enemy
    {
        public Giant(int level) : base("Gigante", level)
        {
            hpScaleFactor = 9;
            attackScaleFactor = 4;
            expOnDeath = 60;
            speed = 1;
            defense = 3;
            ScaleStats();
        }

    }

    public class DarkWizard : Enemy
    {
        public DarkWizard(int level) : base("Mago Oscuro", level)
        {
            hpScaleFactor = 7;
            attackScaleFactor = 6;
            expOnDeath = 70;
            speed = 4;
            defense = 2;
            ScaleStats();
        }
    }
}

File: Java/ProyectoParejasPOOJava/src/Enemy.java
public abstract class Enemy extends Character {

    public int expOnDeath;
    public int hpScaleFactor;
    public int atkScaleFactor;
    public DeathCallBack onDeath;

    public Enemy(String name, int level) {
        super(name);
        this.level = level;
        actions.add(new SingleTargetAttack());
    }

    @Override
    public CharacterAction chooseAction() {
        java.util.Random ran = new java.util.Random();
        int index = ran.nextInt(actions.size());
        CharacterAction action = actions.get(index);
        action.user = this;
        return action;
    }

    public void scaleStats() {
        maxHP      += hpScaleFactor * level;
        hp          = maxHP;
        atk        += atkScaleFactor * level;
        defense    += defense * level;
        expOnDeath += (expOnDeath * level) - level * 2;
    }
}

class Slime extends Enemy {
    public Slime(int level) {
        super("Slime", level);
        hpScaleFactor  = 3;
        atkScaleFactor = 1;
        expOnDeath     = 10;
        scaleStats();
    }
}

class Orc extends Enemy {
    public Orc(int level) {
        super("Orco", level);
        hpScaleFactor  = 4;
        atkScaleFactor = 2;
        expOnDeath     = 20;
        scaleStats();
    }
}

class Troll extends Enemy {
    public Troll(int level) {
        super("Troll", level);
        hpScaleFactor  = 10;
        atkScaleFactor = 2;
        expOnDeath     = 30;
        scaleStats();
    }
}

class MotherSlime extends Enemy {
    public MotherSlime(int level) {
        super("Madre Slime", level);
        hpScaleFactor  = 8;
        atkScaleFactor = 2;
        expOnDeath     = 50;
        scaleStats();
    }
}

class Giant extends Enemy {
    public Giant(int level) {
        super("Gigante", level);
        hpScaleFactor  = 12;
        atkScaleFactor = 5;
        expOnDeath     = 60;
        scaleStats();
    }
}

class DarkWizard extends Enemy {
    public DarkWizard(int level) {
        super("Mago Oscuro", level);
        hpScaleFactor  = 7;
        atkScaleFactor = 6;
        expOnDeath     = 70;
        scaleStats();
    }
}
